feat: blog: Introducing bulk suppressions (#712)
eslint/eslint.org: src/content/drafts/introducing-bulk-suppressions.md

ESLint's new suppression system offers a more flexible approach, allowing you to:

1. Record existing violations in a separate file that can be checked in to source control.
2. Enable the rule as an error. This will now only apply to new code.
3. Address existing violations recorded in the suppressions file at your own pace.

This approach prioritizes preventing new violations while providing a clear path to incrementally improve existing code.

## Getting started

Once you've configured a new rule as `"error"` in your ESLint configuration, you can bulk suppress existing violations using the command line. You can choose to suppress all violations or only those for specific rules.

In either case, it's recommended to run the suppression command with the `--fix` flag to automatically resolve any violations that can be fixed.

```
# Suppress all violations
eslint --suppress-all

> # Fix all violations and suppress them

Reviewer: ```suggestion
# Fix all autofixable violations and suppress the remaining ones
```

Maybe something like this would be clearer?

### Changing the suppressions file location

By default, ESLint creates the suppressions file as `eslint-suppressions.json` in the project root. You can use the `--suppressions-location` option to specify a different location for the suppressions file.

The `--suppressions-location` option accepts a path to the file, which can be either absolute or relative to the current working directory. For example, the following command stores the suppressions in a file called `suppressions.json` in the current working directory:

```
eslint --suppress-all --suppressions-location ./suppressions.json
```

Note that you must specify the `--suppressions-location` option when running ESLint commands that use the suppressions file. For example, to run ESLint with the suppressions file from the example above:

```
eslint --suppressions-location suppressions.json
```

### Managing suppressions over time

After enabling suppressions, it is recommended to commit the suppressions file to your repository. This ensures all team members have the same suppressions active when running ESLint.

> As you fix suppressed violations over time, the suppressions file will become outdated. ESLint outputs a warning when it finds suppressed warnings that no longer occur:

Reviewer: ```suggestion
As you fix suppressed violations over time, the suppressions file will become outdated. ESLint outputs a warning when it finds suppressed violations that no longer occur:
```

To avoid confusion as warning-level lint messages are not suppressed.

---
layout: post
title: Introducing bulk suppressions
teaser: "How to enable stricter linting incrementally"
tags:
  - Suppressions
  - Command Line
authors:
  - softius
categories:
  - Announcements
---

At some point, every development team faces the same challenge: you want to enable a new ESLint rule to improve code quality, but your existing codebase already contains numerous violations. When the rule isn't auto-fixable, the path to enabling it as an error can be challenging.

Today, we're excited to introduce ESLint's new violation suppression system, designed to make adopting stricter linting rules more manageable. Let's explore how it works and how it can help your team enhance code quality incrementally.

## Why bulk suppressions?

When enabling a rule in an established project, teams typically choose between fixing all violations upfront, using warnings instead of errors, disabling rules in specific files, or adding inline disable comments. Each approach has merit, but none provides an ideal balance between enforcing standards for new code while addressing existing violations over time.

> ESLint's new suppression system offers a more flexible approach by:
> 
> 1. Recording existing violations in a separate file
> 2. Enforcing the rule as an error for all new code
> 3. Suppressing recorded violations, allowing you to address existing violations at your own pace

Reviewer 2: ```suggestion
ESLint's new suppression system offers a more flexible approach, allowing you to:

1. Record existing violations in a separate file that can be checked in to source control.
2. Enable the rule as an error. This will now only apply to new code.
3. Address existing violations recorded in the suppressions file at your own pace.
```

This approach prioritizes preventing new violations while providing a clear path to incrementally improve existing code.

## Getting started

Once you've configured a new rule as `"error"` in your ESLint configuration, you can bulk suppress existing violations using the command line. You can choose to suppress all violations or specific rules.

```
# Suppress all violations
eslint --suppress-all

# Suppress specific rules
eslint --suppress-rule <rule-name>
```

When you execute any of the above commands, ESLint will create a file called `eslint-suppressions.json` in the project root. This file contains the number of violations for each rule and file. The suppressions file is structured as follows:

```json
{
  "src/file1.js": {
    "no-undef": {
        "count": 1
    }
  },
  "src/file2.js": {
    "no-unused-expressions": {
        "count": 2
    }
  }
}
```

> ESLint will suppress all `"error"` violations for the specified rules in subsequent runs and will not treat them as errors. However, if more violations are found for the same rule in the same file, ESLint will still report **all** of them in the output. For example, if a file originally had 2 suppressed errors but now has 5, ESLint will display all 5 errors.

Reviewer 2: ```suggestion
ESLint will suppress all `"error"` violations for the specified rules in subsequent runs and will not report them. However, if more violations are found for the same rule in the same file, ESLint will report **all** of them. For example, if a file originally had 2 suppressed `no-undef` errors but now has 5, ESLint will display all 5 `no-undef` errors.
```

> This approach is intentional and helps avoid confusion. Since suppressions are tracked per file and rule—not per specific line or code change—there’s no reliable way to determine whether the new violations were introduced recently or already existed. Rather than trying to guess which violations to hide, ESLint chooses to show the full picture.

Reviewer 2: ```suggestion
This approach is intentional and helps avoid confusion. Because suppressions are tracked per file and rule—not per specific line or code change—there’s no reliable way to determine whether the new violations were introduced recently or already existed. Rather than trying to guess which violations to hide, ESLint chooses to show the full picture.
```

> Please note that it's recommended to combine the above commands with the `--fix` flag to fix any violations that can be automatically resolved.

Reviewer 2: I think we should move this note right under the actual commands. It's a bit out of place here.

### Changing the suppressions file location

By default, ESLint creates the suppressions file as `eslint-suppressions.json` in the project root. You can use the `--suppressions-location` option to specify a different location for the suppressions file. This is useful if you want to store the suppressions in a specific directory.

> Please note when using a custom suppressions location, you must provide the `--suppressions-location` option for every ESLint command, not just when creating the suppressions file.

Reviewer 2: Can we expand on this with a concrete example.

I believe you're trying to say "make sure to include the suppressions file location whenever you lint files," it's just not very clear.

Author: Yes, exactly. I will work on this a bit more.

```
eslint --suppressions-location suppressions.json
```

### Managing suppressions over time

After enabling suppressions, it is recommended to commit the suppressions file to your repository. This ensures all team members have the same suppressions active when running ESLint.

> As you fix suppressed violations over time, the suppressions file will become outdated. When this happens, ESLint will notify you:

Reviewer 2: ```suggestion
As you fix suppressed violations over time, the suppressions file will become outdated. ESLint outputs a warning when it finds suppressed warnings that no longer occur:
```

Once you've configured a new rule as `"error"` in your ESLint configuration, you can bulk suppress existing violations using the command line. You can choose to suppress all violations or specific rules.

Please note that it's recommended to run the suppression command with the `--fix` flag to automatically resolve any violations that can be fixed.

```

> # Suppress all violations

Reviewer 2: Should we remove this line?

Author: I think not because it provides a short description for each command. Keeping it consistent.

Reviewer 2: My concern is that immediately before it we're telling people to use `--fix` and then showing an example where it's not using `--fix`. It seems like one of these two needs to change to be consistent.

Author: Clear. I have removed the examples without --fix

eslint --suppress-all

# Fix all violations and suppress them
eslint --suppress-all --fix

# Suppress specific rules

> eslint --suppress-rule <rule-name>

Reviewer 2: Should we add `--fix` here?

Author: `--suppress-rule <rule-name> --fix` is a bit different since it will fix all the violations, not just for the rule provided. And I didn't want to give the wrong impression. Maybe we can go with something like:

 ```suggestion
eslint --suppress-rule <rule-name>

# Fix all violations, but suppress only violations for <rule-name>
eslint --suppress-rule <rule-name> --fix
```
